File: citygroves/frontend/flesh/src/apis/backend.ts
import axios from 'axios';
import {Config} from '@/config'
import * as url from "url";


class ApiObject {
    api_interface: ApiInterface;
    id: Number;
    str_repr: string;

    /**
     * Patch an object
     * @param {exclude} exclude fields from patching.
     */
    save(exclude?: Array<string>) {
        this.api_interface.update(this, exclude);
    }

    async _getFile(url: string, filename: string) {
        let response = await axios({
            url: url,
            method: 'GET',
            responseType: 'blob',
        });

        var fileURL = window.URL.createObjectURL(new Blob([response.data]));
        var fileLink = document.createElement('a');
        fileLink.href = fileURL;
        fileLink.setAttribute('download', filename);
        document.body.appendChild(fileLink);
        fileLink.click();
    }
}

export class Address extends ApiObject {
    raw_address: string;
    street_line1: string;
    street_line2: string;
    street_line3: string;
    city: string;
    state: string;
    post_code: string;
    country: string;
}

export class Person extends ApiObject {
    first_name: string;
    middle_names: string;
    last_name: string;
    email: string;
    phone: string;
    dob: Date;
}

export class Tenant extends ApiObject {
    person: Person;
    room: Room;
    lease_start: Date;
    lease_end: Date;
    str_repr: string;

    getTitle(): string {
        return `${this.str_repr} U${this.room.unit.number}R${this.room.number}`
    }
}

export class EntryNotice extends ApiObject {
    tenant: Number;
    planned_on: Date;
    planned_time: string;
    is_inspection: Boolean;
    is_cleaning: Boolean;
    is_repairs_or_maintenance: Boolean;
    is_pest_control: Boolean;
    is_showing_to_buyer: Boolean;
    is_valutation: Boolean;
    is_fire_and_rescue: Boolean;
    details: string;

    getPdf() {
        this._getFile(url.resolve(this.api_interface.url, String(`${this.id}/?pdf/`)),
            `${this.str_repr}.pdf`);
    }

    send() {
        axios.post(url.resolve(this.api_interface.url, String(`${this.id}/send/`)));
    }
}

export class Referrer extends ApiObject {
    first_name: string;
    last_name: string;
    email: string;
    phone: string;
    address: Address;
    dob: Date;
}

export class Unit extends ApiObject {
    number: Number
}

export class Room extends ApiObject {
    number: Number;
    unit: Unit;
}


export class Application extends ApiObject {
    api_interface: ApiInterface;

    id: Number;
    person: Person;
    room: Room;
    current_address: Address;
    number_of_ppl_to_move_in: Number;
    move_in_date: Date;
    guarantor_will_pay: Boolean;
    is_employed: Boolean;
    centerlink_will_pay: Number;
    have_sufficient_funds: Boolean;
    is_local_student: Boolean;
    is_international_student: Boolean;
    is_young_professional: Boolean;
    referrers: Array<Referrer>;
    digital_signature: string;

    getTitle(): string {
        return `${this.person.first_name} ${this.person.last_name} U${this.room.unit.number}R${this.room.number}`
    }

    save() {
        super.save(["person", "room", "current_address", "referrers"]);
    }

    getLease() {
        this._getFile(url.resolve(this.api_interface.url, String(`${this.id}/getlease/`)),
            `${this.getTitle().replace(/\s/g, "")}.pdf`);
    }
}

class ApiInterface {
    api: Api;
    ApiObjectType;
    url: string;

    constructor(api: Api, ApiObjectType, endpoint: string) {
        this.api = api;
        this.url = url.resolve(this.api.apiUrl, endpoint);
        this.ApiObjectType = ApiObjectType;
    }

    getAll(): Array<Application> {
        return axios.get(this.url).then(req => {
            let applications: Array<Application> = [];
            for (const obj of req.data) {
                let api_object = new this.ApiObjectType();
                api_object.api_interface = this;

                for (const [name, value] of Object.entries(obj)) {
                    api_object[name] = value;
                }
                applications.push(api_object);
            }
            return applications;
        });
    }

    async get(id: Number) {
        let response = await axios.get(url.resolve(this.url, String(id)));
        let api_object = new this.ApiObjectType();
        api_object.api_interface = this;
        for (const [name, value] of Object.entries(response.data)) {
                    api_object[name] = value;
                }

        return api_object;
    }

    update(apiObject: ApiObject, exclude: Array<string> = []) {
        let payload = Object.assign({}, apiObject);

        // Remove api_interface object so we don't send it to the backend
        delete payload.api_interface;

        for (const excl of exclude) {
            delete payload[excl];
        }

        axios.patch(url.resolve(this.url, String(payload.id) + "/"), payload);
    }

    async create(apiObject: ApiObject) {
        let payload = Object.assign({}, apiObject);
        let response = await axios.post(this.url, payload);
        apiObject.api_interface = this;
        apiObject.id = response.data.id;
        apiObject.str_repr = response.data.str_repr;
    }
}

class Api {
    apiUrl: string;
    applications: ApiInterface;
    people: ApiInterface;
    tenants: ApiInterface;
    addresses: ApiInterface;
    referrer: ApiInterface;
    entry_notices: ApiInterface;

    constructor() {
        this.apiUrl = Config.backend.apiUrl;
        this.applications = new ApiInterface(this, Application, "applications/");
        this.people = new ApiInterface(this, Person, "people/");
        this.tenants = new ApiInterface(this, Tenant, "tenants/");
        this.addresses = new ApiInterface(this, Address, "addresses/");
        this.referrer = new ApiInterface(this, Referrer, "referrers/");
        this.entry_notices = new ApiInterface(this, EntryNotice, "entry-notices/");
    }
}

export let api: Api = new Api();

File: citygroves/comm/python/cg_comm/api_clients/backend.py
import json
from datetime import date
from typing import List, Dict, Any, Optional

from furl import furl
from requests import Session, Response
from requests.adapters import HTTPAdapter

from dataclasses import dataclass, asdict, field
from django.core.serializers.json import DjangoJSONEncoder


@dataclass
class Address:
    raw_address: str
    street_line1: Optional[str] = None
    street_line2: Optional[str] = None
    street_line3: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    post_code: Optional[str] = None
    country: Optional[str] = None


@dataclass
class Person:
    first_name: str
    last_name: str
    email: str
    phone: str
    dob: Optional[date] = None


@dataclass
class Referrer:
    first_name: str
    last_name: str
    email: str
    phone: str
    address: Address
    dob: Optional[date] = None


@dataclass
class Application:
    person: Person
    unit_number: int
    room_number: int
    current_address: Address
    number_of_ppl_to_move_in: int
    move_in_date: date
    guarantor_will_pay: bool
    is_employed: bool
    centerlink_will_pay: bool
    have_sufficient_funds: bool
    is_local_student: bool
    is_international_student: bool
    is_young_professional: bool
    referrers: List[Referrer] = field(default_factory=list)
    digital_signature: Optional[str] = None


class Applications:
    def __init__(self, backend: 'Backend'):
        self._backend = backend
        self._base_url = furl("applications/")

    def create(self, application: Application) -> None:
        payload: Dict[str, Any] = asdict(application)
        self._backend.post(self._base_url, payload)


class Tenants:
    def __init__(self, backend: 'Backend'):
        self._backend = backend
        self._base_url = furl("tenants")


class Backend:
    def __init__(self, api_url: str):
        self.tenants = Tenants(backend=self)
        self.applications = Applications(backend=self)

        self._base_url: furl = furl(api_url)

        self._session = Session()
        self._session.headers = {
            "Content-Type": "application/json; charset=utf-8",
            "Accept": "application/json",
        }

        adapter = HTTPAdapter(max_retries=3)
        self._session.mount("http://", adapter)
        self._session.mount("https://", adapter)

    def post(self, endpoint: furl, data: Dict[str, Any]) -> None:
        url = furl(self._base_url)
        url.add(path=str(endpoint))
        response: Response = self._session.post(url, data=json.dumps(data, cls=DjangoJSONEncoder))
        response.raise_for_status()

    def get(self, endpoint: furl) -> Dict[str, Any]:
        url = furl(self._base_url)
        url.add(path=str(endpoint))
        response: Response = self._session.get(url)
        response.raise_for_status()
        return response.content
